pkg/objstore/bos/bos.go

```diff
@@ -0,0 +1,392 @@
+func (b *Bucket) Upload(_ context.Context, name string, r io.Reader) error {
+	size, err := objstore.TryToGetSize(r)
+	if err != nil {
+		return errors.Wrapf(err, "failed to get size apriori to upload %s", name)
+	}
+
+	partNums, lastSlice := int(math.Floor(float64(size)/partSize)), size%partSize
+	if partNums == 0 {
+		body, err := bce.NewBodyFromSizedReader(r, lastSlice)
+		if err != nil {
+			return errors.Wrap(err, "failed to NewBodyFromSizedReader")
+		}
+
+		if _, err := b.client.PutObject(b.name, name, body, nil); err != nil {
+			return errors.Wrap(err, "failed to upload bos object")
+		}
+
+		return nil
+	}
+
+	result, err := b.client.BasicInitiateMultipartUpload(b.name, name)
+	if err != nil {
+		return errors.Wrap(err, "failed to initiate multi-part upload")
+	}
+
+	uploadEveryPart := func(partSize int64, part int, uploadId string) (string, error) {
+		body, err := bce.NewBodyFromSizedReader(r, partSize)
+		if err != nil {
+			return "", errors.Wrap(err, "failed to NewBodyFromSectionFile")
+		}
+
+		etag, err := b.client.UploadPart(b.name, name, uploadId, part, body, nil)
+		if err != nil {
+			if err := b.client.AbortMultipartUpload(b.name, name, uploadId); err != nil {
+				return etag, errors.Wrap(err, "failed to abort multi-part upload")
+			}
+			return etag, errors.Wrap(err, "failed to upload multi-part part")
+		}
+		return etag, nil
+	}
+
+	var parts []api.UploadInfoType
+
+	for part := 1; part <= partNums; part++ {
+		etag, err := uploadEveryPart(partSize, part, result.UploadId)
+		if err != nil {
+			return errors.Wrap(err, "failed to upload every part")
+		}
+		parts = append(parts, api.UploadInfoType{PartNumber: part, ETag: etag})
+	}
+
+	if lastSlice != 0 {
+		etag, err := uploadEveryPart(lastSlice, partNums+1, result.UploadId)
+		if err != nil {
+			return errors.Wrap(err, "failed to upload the last part")
+		}
+		parts = append(parts, api.UploadInfoType{PartNumber: partNums + 1, ETag: etag})
+	}
+
+	if _, err := b.client.CompleteMultipartUploadFromStruct(b.name, name, result.UploadId, &api.CompleteMultipartUploadArgs{Parts: parts}); err != nil {
+		return errors.Wrap(err, "failed to set multi-part upload completed")
+	}
+	return nil
+}
+
+// Iter calls f for each entry in the given directory (not recursive). The argument to f is the full
+// object name including the prefix of the inspected directory.
+func (b *Bucket) Iter(ctx context.Context, dir string, f func(string) error, opt ...objstore.IterOption) error {
+	if dir != "" {
+		dir = strings.TrimSuffix(dir, objstore.DirDelim) + objstore.DirDelim
+	}
+
+	delimiter := objstore.DirDelim
+
+	if objstore.ApplyIterOptions(opt...).Recursive {
+		delimiter = ""
+	}
+
+	var marker string
+	for {
+		if err := ctx.Err(); err != nil {
+			return errors.Wrap(err, "context closed while iterating bucket")
+		}
+
+		objects, err := b.client.ListObjects(b.name, &api.ListObjectsArgs{
+			Delimiter: delimiter,
+			Marker:    marker,
+			MaxKeys:   1000,
```

Comment:
What if there are more than that? :thinking:

Reply:
> There's no context.Context support? :(

Yes, not supported in current SDK.

Reply:
> What if there are more than that? 🤔

If there are more than 1000, the server will return objects with two fields `objects.NextMarker` and `objects.IsTruncated`, if  !objects.IsTruncated, we continue to ListObjects in the next loop.

Reply:
Can we add at least issue on SDK on ability to cancel request from context? 🤗

Reply:
There is already an issue about this. 😂